### ISSUE TITLE
Dropout of attention weights in function F.multi_head_attention_forward() breaks sum-to-1 constraint

### ISSUE THREAD

**Issue by hantek** [pytorch/pytorch#33360]
## 🐛 Bug

Dropout by calling the built-in dropout function includes rescaling the un-dropped elements, which results in the dropped attention weight vectors possibly sum to a larger than 1 value. 

 
## To Reproduce

See line 3374 to 3376 in file `torch/nn/functional.py`

```
        attn_output_weights = softmax(
            attn_output_weights, dim=-1)
        attn_output_weights = dropout(attn_output_weights, p=dropout_p, training=training)

       attn_output = torch.bmm(attn_output_weights, v)
```

(Pdb) attn_output_weights.sum(2)
tensor([[1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        ...,
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000],
        [1.0000, 1.0000, 1.0000,  ..., 1.0000, 1.0000, 1.0000]],
       device='cuda:0', grad_fn=<SumBackward1>)
(Pdb) n
> /home/hantek/.conda/envs/py37/lib/python3.7/site-packages/torch/nn/functional.py(3378)multi_head_attention_forward()
-> attn_output = torch.bmm(attn_output_weights, v)
(Pdb) attn_output_weights.sum(2)
tensor([[1.0307, 1.0081, 1.0560,  ..., 1.0524, 1.0065, 0.9454],
        [1.0181, 1.0151, 1.0170,  ..., 1.0090, 0.9986, 0.9864],
        [0.9757, 0.9827, 0.9963,  ..., 1.0014, 0.9632, 0.9294],
        ...,
        [1.0049, 0.9946, 0.9487,  ..., 1.0083, 1.0059, 0.9952],
        [0.9971, 1.0024, 1.0114,  ..., 0.9894, 1.0492, 1.0207],
        [1.0050, 1.0084, 1.0159,  ..., 0.9867, 0.9799, 1.0328]],
       device='cuda:0', grad_fn=<SumBackward1>)

What I expect to happen is that the weights still sum-to-1, or sum to a smaller than 1 value after attention dropout.

## Environment
python 3.7, pytorch 1.4.0

**Comment by VitalyFedyunin** [comment]
Hi! Can you please add code how you construct `dropout` in this example. Is it module or imported function?

**Comment by hantek** [comment]
It is pytorch built-in function, i.e., ``torch.nn.functional.dropout()''

See https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L788-L807

**Comment by hantek** [comment]
Basically we need a non-scaling dropout function in cpp backend, and then call that function instead of F.dropout() in all cases that have sum-to-1 constraints or similar constraints.
